dispatch.go

```diff
@@ -327,57 +371,44 @@ func (d *dispatch) dispatchResponse(m *message) (err error) {
 	return
 }
 
-func (d *dispatch) Reset() error {
-	d.callsMutex.Lock()
-	for k, v := range d.calls {
-		v.ch <- EofError{}
-		delete(d.calls, k)
+func (d *dispatch) decodeMessage(m *message, i interface{}) error {
+	err := d.dec.Decode(i)
+	if err == nil {
+		m.nDecoded++
 	}
-	d.callsMutex.Unlock()
-	return nil
+	return err
 }
 
-func (d *dispatch) Dispatch(m *message) (err error) {
-	switch m.nFields {
-	case 3:
-		err = d.dispatchTriple(m)
-	case 4:
-		err = d.dispatchQuad(m)
-	default:
-		err = NewDispatcherError("can only handle message quads (got n=%d fields)", m.nFields)
+func (d *dispatch) decodeToNull(m *message) error {
+	var err error
+	for err == nil && m.nDecoded < m.nFields {
+		i := new(interface{})
+		d.decodeMessage(m, i)
 	}
-	return
+	return err
 }
 
-func (d *dispatch) dispatchTriple(m *message) (err error) {
-	var l int
-	if err = m.Decode(&l); err != nil {
-		return
-	}
-	switch l {
-	case TYPE_NOTIFY:
-		d.dispatchNotify(m)
-	default:
-		err = NewDispatcherError("Unexpected message type=%d; wanted NOTIFY=%d",
-			l, TYPE_NOTIFY)
+func (d *dispatch) decodeError(m *message, f ErrorUnwrapper) (app error, dispatch error) {
+	var s string
+	if f != nil {
+		arg := f.MakeArg()
+		err := d.decodeMessage(m, arg)
+		if err != nil {
+			return nil, err
+		}
+		return f.UnwrapError(arg)
+	} else if dispatch = d.decodeMessage(m, &s); dispatch == nil && len(s) > 0 {
+		app = errors.New(s)
 	}
 	return
 }
 
-func (d *dispatch) dispatchQuad(m *message) (err error) {
-	var l int
-	if err = m.Decode(&l); err != nil {
-		return
-	}
-
-	switch l {
-	case TYPE_CALL:
-		d.dispatchCall(m)
-	case TYPE_RESPONSE:
-		d.dispatchResponse(m)
-	default:
-		err = NewDispatcherError("Unexpected message type=%d; wanted CALL=%d or RESPONSE=%d",
-			l, TYPE_CALL, TYPE_RESPONSE)
+func wrapError(f WrapErrorFunc, e error) interface{} {
+	if f != nil {
```

Comment:
same here

(might be worth doing a govet / golint pass over this code in a separate PR)